buf.md

```diff
@@ -0,0 +1,47 @@
+Byte-Buffer Abstraction
+=======================
+
+```k
+requires "evm.md"
+
+module BUF-SYNTAX
+    imports EVM
+```
+
+Both `#bufStrict(SIZE, DATA)` and `#buf(SIZE, DATA)` represents a symbolic byte array of length `SIZE` bytes, left-padded with zeroes.
+Version `#bufStrict` is partial and only defined when `DATA` is in the range given by `SIZE`.
+It rewrites to `#buf` when data is in range, and is expected to immediately evaluate into `#buf` in all contexts.
+Version `#buf` is total and artificially defined `modulo 2 ^Int (SIZE *Int 8)`.
+This division is required to facilitate symbolic reasoning in Haskell backend, because Haskell has limitations
+when dealing with partial functions.
+
+**Usage:** All symbolic byte arrays must be originally created as `#bufStrict`.
+This ensures `#buf` is never present in out of range mode.
+For this, definition rule RHS should always use `#bufStrict` when array is first created, but may use `#buf` when array
+is just carried on from LHS without changes. Definition rule LHS should only use `#buf`.
+Claims should always use `#bufStrict` in LHS and `#buf` in RHS.
+
+```k
+    syntax ByteArray ::= #bufStrict ( Int , Int ) [function]
+    syntax ByteArray ::= #buf ( Int , Int ) [function, functional, smtlib(buf)]
+
+endmodule
+
+module BUF
+    imports BUF-SYNTAX
+
+    rule #bufStrict(SIZE, DATA) => #buf(SIZE, DATA)
+      requires #range(0 <= DATA < (2 ^Int (SIZE *Int 8)))
+
+    rule #buf(SIZE, DATA) => #padToWidth(SIZE, #asByteStack(DATA %Int (2 ^Int (SIZE *Int 8))))
+      [concrete]
+
+endmodule
+
+module BUF-KORE [kore, symbolic]
+    imports BUF
```

Comment:
The import relation must be opposite: BUF must import BUF-KORE. This is the pattern used in domains.k . Otherwise any module that needs BUF would also have to import BUF-KORE.

Reply:
The rest looks good.

Reply:
The problem is that when compiling with the LLVM backend (which is needed for including `abi.md` into definitions like Firefly), then you have to say `imports BUF`, which ignores the `BUF-KORE` module, and then ignores the `BUF-SYNTAX` module, so then it complains that `#buf` and `#bufStrict` do not exist. Another solution is to have `BUF` import both `BUF-KORE` and `BUF-SYNTAX`, if you're OK with that..

Reply:
Yes, that' what I meant. Relation must be BUF imports BUF-KORE,BUF-SYNTAX ; BUF-KORE imports BUF-SYNTAX.